README.md
# io_import_x
(based on directX_blender by littleneo)
A DirectX importer addon for Blender 2.7

## Primary Goals

* Import anything from an .x file
  (obviously verts and faces but also uv, armatures, weights, normals).
* Import .x in binary format [not yet implemented]

-littleneo and Poikilos

## Setup
1. Start Blender
2. Download & *Save* zip from GitHub. If you click open you may not be
   able to find the zip later _(you must use the poikilos fork or
   accepted git commits in order for this method to work, otherwise copy
   the extracted io_directx_bel folder to
   ~/.config/blender/2.79/scripts/addons/ on GNU+Linux systems or
   %APPDATA%/Blender Foundation/blender/2.79/scripts/addons)_ on Windows
   and skip to step 3)
   * Click "File," "User Preferences," "Add-ons"
   * Click "Install add-on from file..."
   * Choose the downloaded io_import_x-master.zip (usually in your
     "Downloads" directory)
3. Enable the add-on in the "Add-ons" tab of "User Preferences".
4. Click "Save User Preferences" to keep the add-on enabled for other
   scenes.

-Poikilos

## Usage
* File > Import > DirectX

## Planned Features
* Export to .x or mod/co-maintain the existing x exporter.
* This project is also a prototype for a 'Blender Exchange Layer'
  project. BEL would be a common layer logically located between an
  importer/exporter addon and the blender data format, that would
  provide a common set of...:
  * methods to retrieve/inject objects in Blender
  * transformation and selection tools between an import/export script
    and Blender (rotate, rescale, filter, etc.)
  * parsing helpers for new io addons
* PLY won't be used unfortunately (it's way too slow as far as I tested)
-littleneo and Poikilos
[See also TODO below]

## Changes
see CHANGELOG.md

## Developer Notes
* I don't want to load the whole file into memory as it can be huge, so
  the importer processes the file in chunks.
* I want random access to 3D data to import pieces, not always
  everything:
  1. Preprocess entire file, retrieving tokens and building an empty
     internal dict with only pointers, no 3D data.
  2. Call any token by name and retrieve the 3D data thanks to pointers
     stored in dicts.
-littleneo and Poikilos

### TODO
* Between step 1 and 2 above, a UI should be provided to select,
  transform, or otherwise process before import.
* I need to know the pointer position of tokens but data.tell() is slow
  a += pointer computed from line length is way faster. To deal with
  EOL (end of line), use rb (read binary) mode; `readline()` is ok in
  binary mode 'rb' with \r\n (win) \n (unix) but not \r mac.
  - Reads 2 characters for windows
  - Reads 1 character if file was produced on mac/Linux
  - On Linux, a Windows EOL (\r\n) becomes \n\n (adding an extra line)
  - On Mac, the \r in a Windows EOL becomes \n so line numbering and
    data read is wrong.
  - Catching this in binary mode allows support for bad newlines (mixed
    \r and \r\n)
  - For now it only works for text format, but the functions called will
    be independent of the container type.

-littleneo and Poikilos


### File Format Parsing

#### Read Main Structure

##### Read Main Token Names
(any 'template', any 'frame', any 'mesh')
* store names in a token directory:
  * token['template'] for templates:
    ```
token['template'][templatename]
token['template'][templatename]['pointer']          (int) chr position in .x file (tell() like*)
token['template'][templatename]['line']             (int) line number in .x file
```
  * token['frame'] for frame and mesh type:
    ```
token['template'][frame or mesh name]
token['template'][frame or mesh name]['pointer']    (int) chr position in .x file (tell() like*)
token['template'][frame or mesh name]['line']       (int) line number in .x file
token['template'][frame or mesh name]['type']       (str) 'ob/bone' or 'mesh'
token['template'][frame or mesh name]['parent']     (str) frame parent of current item
token['template'][frame or mesh name]['childs']     (str list) list of child frame or mesh names
token['template'][frame or mesh name]['matrix']     (int) for now chr position of FrameTransformMatrix
```

At the end of main structure, the script prints a tree of the data.

#### Read Template Definitions

For each template in the dict, populate definitions in it.
Create new fields in each token['template'][templatename]
according to values found in .x:
```
token['template'][templatename]['uuid']                 (str) <universally unique identifier>
token['template'][templatename]['members']['name']      (str) member name
token['template'][templatename]['members']['type']      (str) DWORD,FLOAT etc keywords or template name
token['template'][templatename]['restriction']          (str) 'open' , 'closed' , or the specidied (restricted) value
```

### Optimization
Allow 2 steps: preprocess, then random access to file:
* First, preprocess the file quickly. Only retrieve main info--nothing
  about verts, faces etc info like number of mats, textures,
  objects/mesh/bone trees: 150000 lines in 5 secs
* Allow user to select what to import.
* Retrieve selected data, using the 'pointer' value to seek() to the
  needed data, then grab/parse/translate into something usable.
* Templates are used at this point to know how to parse individual parts
  (the parser is adaptive).

So far this looks fast (tested on windows).
Preprocessing can be important because of eol and the code I wrote to
compute pointer value.
* (data.tell() is slow)
* Only one .x file tested, header is: xof 0303txt 0032 (windows \r\n eol)
* I don't know a lot about .x format:
  * **uuid** :
    are the member/restriction always the same for a same uuid/template ?
    template name can vary for a same uuid ?
  * **syntax** :
    blank lines IN a stream of a {} section, after ; ?
    comments // and # IN a stream of data ?
    '{' and '<something>' and '}' on the same line or '{' '}' are always unique ?
-littleneo

## Credits
* TEST FILES:
  <http://assimp.svn.sourceforge.net/viewvc/assimp/trunk/test/models/X/>

## References
* <http://paulbourke.net/dataformats/directx/>
* <http://www.informikon.com/various/the-simplest-skeletal-animation-possible.html>
* <http://msdn.microsoft.com/en-us/library/windows/desktop/bb173011%28v=VS.85%29.aspx>
* <http://www.toymaker.info/Games/index.html>

## Discussions
* <https://blender.community/c/rightclickselect/kYcbbc/>
* <https://www.reddit.com/r/blender/comments/beq3al/importing_x_files_directx_is_almost_working_help/>
* <https://stackoverflow.com/questions/5767634/import-directx-x-into-blender#comment98211206_16675435>
* <https://blender.stackexchange.com/questions/107834/blender-2-79a-x-files>
* <https://github.com/littleneo/directX_blender/issues/6>
* <https://github.com/limemidolin/directX_blender/pull/1/commits>

## See also
* A fault-tolerant x loader in Irrlicht:
  <https://sourceforge.net/p/irrlicht/code/HEAD/tree/trunk/source/Irrlicht/CXMeshFileLoader.cpp#l740>
* The downstream version in the Blender repository
  <https://svn.blender.org/svnroot/bf-extensions/contrib/py/scripts/addons/io_directx_bel/__init__.py>
  is missing fixes from the <2013 littleneo version, so I mentioned that
  as a comment to my Right-Click Select post at the link under
  "Discussions" above. Also, they seemed to have done a search and
  replace do update certain things for Blender 2.8 but it is not working
  in either 2.79b or 2.80 (nightly 2018-04-18).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on
[Keep a Changelog](https://keepachangelog.com/en/1.0.0/).


## [git] - 2021-11-05
### Fixed
- Fix uses of Blender API that are deprecated as of 2.8: Search for
  every "Removed" or "Renamed" feature in the entire [2.8
  changelog](https://docs.blender.org/api/blender2.8/change_log.html)
  other than `bpy.types.SpaceView3D`, `ThemeClipEditor`, `Theme*`
  `ToolSettings`, `UILayout`, `WorldLighting`, `GPencil`,
  `ParticleSettings`, `Text`, `SPHFluidSettings`, and other scene- or
  render-related things that didn't seem to apply to io_import_x.


## [git] - 2021-03-31
### Changed
- Update for Blender 2.80
  - Merge egore changes
  - Add some instances of `if bpy.app.version >= (2, 80, 0):` (in some or all cases listed below)
  - Change uv_textures to uv_layers: <https://blender.stackexchange.com/questions/193894/updating-blender-2-79-script-to-2-80-mesh-uv-textures-new-fails-with-mesh-o>.
  - `name` must be a keyword argument in vertex_groups.new(...).
  - Use `bpy.context.collection.objects` instead of `bpy.context.scene.objects`.
  - Change `.select = True` to `.select_set(state=True)` (fix "AttributeError: 'Object' object has no attribute 'select'").
  - Change `bpy.context.collection.objects.active = arm` to `bpy.context.view_layer.objects.active = arm` (fix "AttributeError: bpy_prop_collection: attribute "active" not found").
  - Add 1.0 alpha to the color (the x format alpha value, which becomes part of the material settings) to
    fix: ```
  File "/home/owner/.config/blender/2.80/scripts/addons/io_import_x/import_x.py", line 856, in getMesh
    mat.diffuse_color = diffuse_color
ValueError: bpy_struct: item.attr = val: sequences of dimension 0 should contain 4 items, not 3
```

### Fixed
- Don't redefine Python 3 builtin file (change the variable name to file_).
- Add the missing arg to import_dXtree (hmm...try tokenname) to fix "expected string got int"
  - and set `lvl` by name since it is a keyword argument.

## [0.27] - 2019-04-18
(Poikilos)
### Changed
* bump version to 2.7 to indicate it works with blender 2.7x as per
  semantic versioning (API changes with every "tenth" of version number
  of blender).
* warning in `__init__.py` about unmaintained upstream littleneo version
* moved files around so "Install Add-on from File" feature will work
  with zip downloaded from GitHub:
  - removed duplicate directX_blender/README which was exactly the same
    as directX_blender/io_directx_bel/README)
  - moved io_directx_bel files to root of repo
  - renamed repo to directxblender (formerly directX_blender) so it is
    using a PEP8 module name
  - changed py files to reflect new module naming (from io_directx_bel
    to directxblender)
* moved Changelog text from README to CHANGELOG.md; changed
  formatting to markdown; corrected some spelling/grammar mistakes;
  improved wording.
* renamed README to README.md; changed formatting to markdown; corrected
  some spelling/grammar mistakes; improved wording
* fix incorrect "blender" property in bl_info (minor revision should be
  only in second part of tuple, not 2nd and 3rd--third should be 0--see
  <https://wiki.blender.org/wiki/Process/Addons/Guidelines/metainfo>).
  Also update required version to 2.66 since that seems to be when
  compatibility broke according to JLouis-B >3 yr old unaccepted pull
  request to littleneo's repo.
* add entry for limemidolin below (list items are commit names)
* add nohup.out to .gitignore

## [git] - 2018-04-12
(limemidolin)
* Reformat code
* Add feature: Do not add unused material if wanted
* Add feature: Load normals file if exist
* Multiple file import
* Fix to work on Blender 2.79

## [0.18] - 2012-01-25
(last littleneo entry)
* code rewrite according to API bmesh changes (me.polygon and UV layers
  essentially)
* implemented `foreachset()` call for UV import (faster)

## [0.17] - 2012-01-25
* faster (60% faster in some cases): various loop improvements; infile
  template parsing is disabled by default. I saw another bottleneck
  regarding data chunks as string but will wait for binary support for
  better point of view.
* interface cosmetics

## [rc 0.16] - 2012-01-23
* committed to SVN (and littleneo git as usual)
* corrected a bug regarding referenced token parenting
* corrected a bug regarding non-parented meshes
* Option to import armatures/empties is enabled by default.
* Last run importer options are saved in a 'last_run' preset,
  so it can be replayed or saved under another name once a
  particular .x profile has been defined.
* tagged script for 2.6.1

## [rc 0.15] - 2012-01-12
* name conversion changed from 5 to 4 digits
* matname, imagename and texturename fixes:
  * A path test is made at image import time with any existing data
    images, so the same file can't be loaded twice regardless what
    naming method is used (/ or \, rel or abs, etc).
    (`bel.material.new()` and after line 835)
  * Image and texture names should be OK now (tested with
    incrediblylongname.x).
  * Materials are replaced accordingly in existing objects when using
    the 'replace' naming method.
* FYI, the x exporter has the following inconveniences:
  * split linked faces into individual faces
  * inversed uvmapping (y axis)?
    - see testfiles/blender_x_export/incrediblylongname.x

## [git] - 2011-12-31
(and 2011-12-29)
* Cosmetics, code cleaning and optimizations
* bpy.ops.object.select_name methods replaced with
    ob.select = True
    bpy.context.scene.objects.active = ob
* corrected a big bug regarding tokens info appending in dXtree()
* new `bpyname()` method in bel module. removed bel.common

## [git] - 2011-12-26
* armature import and bone max. length option

## [git] - 2011-11-23
* contrib candidate :)
* solved some naming cases, added bel methods.
* added experimental option regarding parenting (no armature yet just
  empties, when needed)
* a bit faster

__init__.py
# Blender DirectX importer
import os
import sys
bl_info = {
    "name": "DirectX Importer",
    "description": "Import DirectX Model Format (.x)",
    "author": "Littleneo (Jerome Mahieux), limemidolin, poikilos",
    "version": (2, 7),
    "blender": (2, 80, 0),
    "location": "File > Import > DirectX (.x)",
    "warning": "Upstream littleneo version does NOT work with >=2.66",
    "wiki_url": "https://github.com/littleneo/directX_blender/wiki",
    "tracker_url": "https://github.com/littleneo/directX_blender/issues",
    "support": "COMMUNITY",  # COMMUNITY is default anyway though
    "category": "Import-Export"
    # I commented undocumented settings. -Poikilos
    # "api": 42615,
    # "dependencies": "",
}

if "bpy" in locals():
    import importlib

    if "import_x" in locals():
        importlib.reload(import_x)
    # if "export_x" in locals():
    #    importlib.reload(export_x)

import bpy
from bpy.props import (BoolProperty,
                       FloatProperty,
                       StringProperty,
                       EnumProperty,
                       CollectionProperty
                       )
from bpy_extras.io_utils import (ExportHelper,
                                 ImportHelper,
                                 path_reference_mode,
                                 axis_conversion,
                                 )

# AttributeError: '_RestrictData' object has no attribute 'filepath':
# myPath = os.path.dirname(bpy.data.filepath)
script_file = os.path.realpath(__file__)
directory = os.path.dirname(script_file)
if not directory in sys.path:
    sys.path.append(directory)

import bel
import bel.mesh
import bel.image
import bel.uv
import bel.material
import bel.ob
import bel.fs


class ImportX(bpy.types.Operator, ImportHelper):
    '''Load a Direct x File'''
    bl_idname = "import_scene.x"
    bl_label = "Import X"
    bl_options = {'PRESET', 'UNDO'}

    files = CollectionProperty(type=bpy.types.PropertyGroup)

    filename_ext = ".x"
    filter_glob = StringProperty(
        default="*.x",
        options={'HIDDEN'},
    )
    show_tree = BoolProperty(
        name="Show x tokens tree",
        description="display relationships between x items in the console",
        default=False,
    )
    show_templates = BoolProperty(
        name="Show x templates",
        description="display templates defined in the .x file",
        default=False,
    )
    show_geninfo = BoolProperty(
        name="Show processing",
        description="display details for each imported thing",
        default=False,
    )
    do_not_add_unused_material = BoolProperty(
        name="Don't add unused materials",
        description="Do not add unused materials.",
        default=False,
    )

    quickmode = BoolProperty(
        name="Quick mode",
        description="only retrieve mesh basics",
        default=False,
    )

    parented = BoolProperty(
        name="Object Relationships",
        description="import armatures, empties, rebuild parent-childs relations",
        default=True,
    )

    bone_maxlength = FloatProperty(
        name="Bone length",
        description="Bones max length",
        min=0.1, max=10.0,
        soft_min=0.1, soft_max=10.0,
        default=1.0,
    )

    chunksize = EnumProperty(
        name="Chunksize",
        items=(('0', "all", ""),
               ('4096', "4KB", ""),
               ('2048', "2KB", ""),
               ('1024', "1KB", ""),
               ),
        default='2048',
        description="number of bytes red in a row",
    )
    naming_method = EnumProperty(
        name="Naming method",
        items=(('0', "increment name if exists", "blender default"),
               ('1', "use existing", "this only append new elements"),
               ('2', "rename existing", "names are forced"),
               ('3', "replace existing", ""),
               ),
        default='0',
        description="behaviour when a name already exists in Blender Data",
    )
    use_ngons = BoolProperty(
        name="NGons",
        description="Import faces with more than 4 verts as fgons",
        default=True,
    )
    use_edges = BoolProperty(
        name="Lines",
        description="Import lines and faces with 2 verts as edge",
        default=True,
    )
    use_smooth_groups = BoolProperty(
        name="Smooth Groups",
        description="Surround smooth groups by sharp edges",
        default=True,
    )

    use_split_objects = BoolProperty(
        name="Object",
        description="Import OBJ Objects into Blender Objects",
        default=True,
    )
    use_split_groups = BoolProperty(
        name="Group",
        description="Import OBJ Groups into Blender Objects",
        default=True,
    )

    use_groups_as_vgroups = BoolProperty(
        name="Poly Groups",
        description="Import OBJ groups as vertex groups",
        default=False,
    )

    use_image_search = BoolProperty(
        name="Image Search",
        description="Search subdirs for any assosiated images " \
                    "(Warning, may be slow)",
        default=True,
    )

    split_mode = EnumProperty(
        name="Split",
        items=(('ON', "Split", "Split geometry, omits unused verts"),
               ('OFF', "Keep Vert Order", "Keep vertex order from file"),
               ),
    )

    global_clamp_size = FloatProperty(
        name="Clamp Scale",
        description="Clamp the size to this maximum (Zero to Disable)",
        min=0.0, max=1000.0,
        soft_min=0.0, soft_max=1000.0,
        default=0.0,
    )

    axis_forward = EnumProperty(
        name="Forward",
        items=(('X', "Left (x)", ""),
               ('Y', "Same (y)", ""),
               ('Z', "Bottom (z)", ""),
               ('-X', "Right (-x)", ""),
               ('-Y', "Back (-y)", ""),
               ('-Z', "Up (-z)", ""),
               ),
        default='-Z',
    )

    axis_up = EnumProperty(
        name="Up",
        items=(('X', "Right (x)", ""),
               ('Y', "Back (y)", ""),
               ('Z', "Same (z)", ""),
               ('-X', "Left (-x)", ""),
               ('-Y', "Front (-y)", ""),
               ('-Z', "Bottom (-z)", ""),
               ),
        default='Y',
    )

    def execute(self, context):
        from . import import_x
        if self.split_mode == 'OFF':
            self.use_split_objects = False
            self.use_split_groups = False
        else:
            self.use_groups_as_vgroups = False

        keywords = self.as_keywords(ignore=("axis_forward",
                                            "axis_up",
                                            "filter_glob",
                                            "split_mode",
                                            ))

        keywords["naming_method"] = int(self.naming_method)

        global_matrix = axis_conversion(from_forward=self.axis_forward,
                                        from_up=self.axis_up,
                                        ).to_4x4()
        keywords["global_matrix"] = global_matrix

        bel.fs.saveOptions(self, 'import_scene.x', self.as_keywords(ignore=(
            "filter_glob",
            "filepath",
            "files"
        )))
        return import_x.load(self, context, **keywords)

    def draw(self, context):
        layout = self.layout

        # import box
        box = layout.box()
        col = box.column(align=True)
        col.label(text='Import Options :')
        col.prop(self, "chunksize")
        col.prop(self, "use_smooth_groups")
        actif = not (self.quickmode)
        row = col.row()
        row.enabled = actif
        row.prop(self, "parented")
        if self.parented:
            row = col.row()
            row.enabled = actif
            row.prop(self, "bone_maxlength")
        col.prop(self, "quickmode")

        # source orientation box
        box = layout.box()
        col = box.column(align=True)
        col.label(text='Source Orientation :')
        col.prop(self, "axis_forward")
        col.prop(self, "axis_up")

        box = layout.box()
        col = box.column(align=True)
        col.label(text="Option 1")
        col.prop(self, "do_not_add_unused_material")

        # naming methods box
        box = layout.box()
        col = box.column(align=True)
        col.label(text='Naming Method :')
        col.props_enum(self, "naming_method")

        # info/debug box
        box = layout.box()
        col = box.column(align=True)
        col.label(text='Info / Debug :')
        col.prop(self, "show_tree")
        col.prop(self, "show_templates")
        col.prop(self, "show_geninfo")

        # row = layout.row(align=True)
        # row.prop(self, "use_ngons")
        # row.prop(self, "use_edges")

        '''
        box = layout.box()
        row = box.row()
        row.prop(self, "split_mode", expand=True)

        row = box.row()
        if self.split_mode == 'ON':
            row.label(text="Split by:")
            row.prop(self, "use_split_objects")
            row.prop(self, "use_split_groups")
        else:
            row.prop(self, "use_groups_as_vgroups")

        row = layout.split(percentage=0.67)
        row.prop(self, "global_clamp_size")

        layout.prop(self, "use_image_search")
        '''


def menu_func_import(self, context):
    self.layout.operator(ImportX.bl_idname, text="DirectX (.x)")


# def menu_func_export(self, context):
#    self.layout.operator(ExportX.bl_idname, text="DirectX (.x)")

def register():
    bpy.utils.register_class(ImportX)

    bpy.types.TOPBAR_MT_file_import.append(menu_func_import)
    # bpy.types.INFO_MT_file_export.append(menu_func_export)


def unregister():
    bpy.utils.unregister_class(ImportX)

    bpy.types.TOPBAR_MT_file_import.remove(menu_func_import)
    # bpy.types.INFO_MT_file_export.remove(menu_func_export)


if __name__ == "__main__":
    register()
